experimental/canvas/graphql/graphql.py

```diff
@@ -0,0 +1,192 @@
+            # added to the GraphQL query, but didn't look it up.
+            # For comparison between call types, just get all enrollments
+            #
+            # TYPE_TEACHER = "TeacherEnrollment"
+            # TYPE_STUDENT = "StudentEnrollment"
+            # if enroll_j["node"]["type"] == TYPE_TEACHER:
+            #     continue
+
+            # Regular API calls default to returning only students in the
+            # "active" and "invited" states.
+            # https://canvas.instructure.com/doc/api/enrollments.html GraphQL
+            # doesn't provide that. For best comparison of results, manually
+            # filter on those.
+            if enroll_j["state"] not in ("active", "invited"):
+                continue
+
+            user_j = enroll_j["user"]
+            students.append(
+                {
+                    "Email": user_j["email"],
+                    "SisIdentifier": user_j["sisId"],
+                    "StudentName": user_j["name"],
+                    "SourceSystemIdentifier": user_j["_id"],
+                }
+            )
+
+            enrollments.append(
+                {
+                    "SectionSourceSystemIdentifier": enroll_j["section"]["_id"],
+                    "StudentSourceSystemIdentifier": user_j["_id"],
+                    "SourceSystemIdentifier": enroll_j["_id"],
+                }
+            )
+
+            # TODO: handle enrollment paging. No idea how to do that right now. Nested paging?
+
+    # Course-level paging
+    course_page = course_connections["pageInfo"]
+    if course_page["hasNextPage"] == True:
+        after = course_page["endCursor"]
+        query = build_course_query(after)
+
+        # Recursive call. This code is quick-and-dirty. Not thinking
+        # very carefully about state and output. Just get it done.
+        post(query)
+
+
+post(build_course_query())
+
+courses_df = pd.DataFrame(courses)
+courses_df.astype({"SourceSystemIdentifier": int})
+courses_df.sort_values(by=["SourceSystemIdentifier"], inplace=True)
+courses_df.to_csv("courses.csv", index=False)
+
+enrollments_df = pd.DataFrame(enrollments)
+enrollments_df.astype({"SourceSystemIdentifier": int})
+enrollments_df.sort_values(by=["SourceSystemIdentifier"], inplace=True)
+enrollments_df.to_csv("enrollments.csv", index=False)
+
+# In this quick-and-dirty, students enrolled in multiple classes will have
+# multiple records. Real solution should de-duplicate.
+students_df = pd.DataFrame(students)
+students_df.astype({"SourceSystemIdentifier": int})
+students_df.sort_values(by=["SourceSystemIdentifier"], inplace=True)
+students_df.to_csv("students.csv", index=False)
+
+elapsed = timeit.default_timer() - start_time
+
+print(f"Elapsed time: {elapsed}")
+
+# Run 1: 11 s
+# Run 2: 7 s
+# Run 3: 7 s
+# Run 4: 7 s
+# Run 5: 6 s
+# Avg: 7 s
+
+# Always takes longer first execution. Probably caching some results. So first execution counts for more.
```

Comment:
This makes sense. They can't possibly index _everything_ to support ad-hoc querying.